In [3]:
import pandas as pd
pd.__version__
import numpy as np

In [4]:
import seaborn as sns

In [5]:
pd.options.display.float_format = '{:,.3f}'.format

#para restringier
# el numero de decimales que se muestra en el dataframe

In [6]:
df = sns.load_dataset('tips')
df

total_bill   tip     sex smoker   day    time  size
0        16.990 1.010  Female     No   Sun  Dinner     2
1        10.340 1.660    Male     No   Sun  Dinner     3
2        21.010 3.500    Male     No   Sun  Dinner     3
3        23.680 3.310    Male     No   Sun  Dinner     2
4        24.590 3.610  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239      29.030 5.920    Male     No   Sat  Dinner     3
240      27.180 2.000  Female    Yes   Sat  Dinner     2
241      22.670 2.000    Male    Yes   Sat  Dinner     2
242      17.820 1.750    Male     No   Sat  Dinner     2
243      18.780 3.000  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [ ]:
#conocer nuestros datos

df.describe(include='all')


total_bill     tip   sex smoker  day    time    size
count      244.000 244.000   244    244  244     244 244.000
unique         nan     nan     2      2    4       2     nan
top            nan     nan  Male     No  Sat  Dinner     nan
freq           nan     nan   157    151   87     176     nan
mean        19.786   2.998   NaN    NaN  NaN     NaN   2.570
std          8.902   1.384   NaN    NaN  NaN     NaN   0.951
min          3.070   1.000   NaN    NaN  NaN     NaN   1.000
25%         13.348   2.000   NaN    NaN  NaN     NaN   2.000
50%         17.795   2.900   NaN    NaN  NaN     NaN   2.000
75%         24.127   3.562   NaN    NaN  NaN     NaN   3.000
max         50.810  10.000   NaN    NaN  NaN     NaN   6.000

In [ ]:
df.groupby(['sex']).aggregate(['min',np.mean,max])

total_bill                 tip              size          
              min   mean    max   min  mean    max  min  mean max
sex                                                              
Male        7.250 20.744 50.810 1.000 3.090 10.000    1 2.631   6
Female      3.070 18.057 44.300 1.000 2.833  6.500    1 2.460   6

In [ ]:
df.groupby(['sex','day'])['total_bill'].aggregate(['min',np.mean,max])

min   mean    max
sex    day                     
Male   Thur 7.510 18.715 41.190
       Fri  8.580 19.857 40.170
       Sat  7.740 20.803 50.810
       Sun  7.250 21.887 48.170
Female Thur 8.350 16.715 43.110
       Fri  5.750 14.146 22.750
       Sat  3.070 19.680 44.300
       Sun  9.600 19.872 35.260

In [ ]:
df['day'].value_counts()

Sat     87
Sun     76
Thur    62
Fri     19
Name: day, dtype: int64

In [ ]:
#si quiero ver la distribucion anterior como un porcentaje 
df['day'].value_counts() /df['day'].value_counts().sum() *100

Sat    35.656
Sun    31.148
Thur   25.410
Fri     7.787
Name: day, dtype: float64

In [ ]:
# ahora me gustaria tener estas estimaciones estadisticas pero por
# alguna de mis categorias

df.groupby('sex').mean()

total_bill   tip  size
sex                           
Male        20.744 3.090 2.631
Female      18.057 2.833 2.460

In [ ]:
# vamos a crear una nueva columna que se va a llamar porcentaje de propina
#que es la relacion entre la propina y el valor de la factura

df['prct_tip']=df['tip']/df['total_bill']
df

total_bill   tip     sex smoker   day    time  size  prct_tip
0        16.990 1.010  Female     No   Sun  Dinner     2     0.059
1        10.340 1.660    Male     No   Sun  Dinner     3     0.161
2        21.010 3.500    Male     No   Sun  Dinner     3     0.167
3        23.680 3.310    Male     No   Sun  Dinner     2     0.140
4        24.590 3.610  Female     No   Sun  Dinner     4     0.147
..          ...   ...     ...    ...   ...     ...   ...       ...
239      29.030 5.920    Male     No   Sat  Dinner     3     0.204
240      27.180 2.000  Female    Yes   Sat  Dinner     2     0.074
241      22.670 2.000    Male    Yes   Sat  Dinner     2     0.088
242      17.820 1.750    Male     No   Sat  Dinner     2     0.098
243      18.780 3.000  Female     No  Thur  Dinner     2     0.160

[244 rows x 8 columns]

In [ ]:
df.groupby('sex').median()

total_bill   tip  size  prct_tip
sex                                     
Male        18.350 3.000     2     0.153
Female      16.400 2.750     2     0.156

In [ ]:
df.groupby('sex').mean()

total_bill   tip  size  prct_tip
sex                                     
Male        20.744 3.090 2.631     0.158
Female      18.057 2.833 2.460     0.166

In [ ]:
df.groupby('sex')[['prct_tip']].describe()

prct_tip                                          
          count  mean   std   min   25%   50%   75%   max
sex                                                      
Male    157.000 0.158 0.065 0.036 0.121 0.153 0.186 0.710
Female   87.000 0.166 0.054 0.056 0.140 0.156 0.194 0.417

In [ ]:
df.groupby('sex')[['total_bill','prct_tip']].describe()

total_bill                            ... prct_tip                        
            count   mean   std   min    25%  ...      min   25%   50%   75%   max
sex                                          ...                                 
Male      157.000 20.744 9.246 7.250 14.000  ...    0.036 0.121 0.153 0.186 0.710
Female     87.000 18.057 8.009 3.070 12.750  ...    0.056 0.140 0.156 0.194 0.417

[2 rows x 16 columns]

# Vamos a hacer una funcion para tronsformar  Euros a dolares (asumimos que el valor de la tabla esta en euros)

In [ ]:
def mean_euro2usd(x):  
  return np.mean(x)*1.12


In [ ]:
mean_euro2usd([100,102])

113.12

In [ ]:
df.groupby('sex')[['total_bill','prct_tip']].apply(mean_euro2usd)

total_bill  prct_tip
sex                         
Male        23.233     0.177
Female      20.224     0.186

In [ ]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].apply(mean_euro2usd)

total_bill  prct_tip
sex    time                        
Male   Lunch       20.214     0.186
       Dinner      24.037     0.174
Female Lunch       18.300     0.182
       Dinner      21.519     0.190

# Para calcular la desviacion estandar

In [ ]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].apply(np.std)

total_bill  prct_tip
sex    time                        
Male   Lunch        7.832     0.045
       Dinner       9.423     0.069
Female Lunch        7.393     0.034
       Dinner       8.123     0.063

# para definir funciones  internamente

osea hacer esto
```
def mean_euro2usd(x):  
  return np.mean(x)*1.12
  
```

In [ ]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].apply(lambda x: np.mean(x*1.12) )

total_bill  prct_tip
sex    time                        
Male   Lunch       20.214     0.186
       Dinner      24.037     0.174
Female Lunch       18.300     0.182
       Dinner      21.519     0.190

# si quiero aplicar varias funciones uso **aggregate** 

tambien funciona resumiendo aggregate como agg

In [ ]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].aggregate([np.mean,np.max])

total_bill        prct_tip      
                    mean   amax     mean  amax
sex    time                                   
Male   Lunch      18.048 41.190    0.166 0.266
       Dinner     21.461 50.810    0.155 0.710
Female Lunch      16.339 43.110    0.162 0.259
       Dinner     19.213 44.300    0.169 0.417

In [ ]:
df.groupby(['sex','time'])[['total_bill','prct_tip']].agg([np.mean,np.max])

total_bill        prct_tip      
                    mean   amax     mean  amax
sex    time                                   
Male   Lunch      18.048 41.190    0.166 0.266
       Dinner     21.461 50.810    0.155 0.710
Female Lunch      16.339 43.110    0.162 0.259
       Dinner     19.213 44.300    0.169 0.417

# Vamos a crear un diccionario de operaciones

---

este diccionario va a contener dos llaves , la primera llave es para las propina y vamos a calcular la propina maxima y miniama

y para el totalbill voy a calcular el promedio y  el valor en moneda dolar




In [ ]:
dict_agg ={'tip':[min,max],'total_bill':[np.mean,mean_euro2usd]}
dict_agg

{'tip': [<function min>, <function max>],
 'total_bill': [<function numpy.mean>, <function __main__.mean_euro2usd>]}

In [ ]:
df.groupby(['sex','time'])[['total_bill','tip']].agg(dict_agg)

tip        total_bill              
                min    max       mean mean_euro2usd
sex    time                                        
Male   Lunch  1.440  6.700     18.048        20.214
       Dinner 1.000 10.000     21.461        24.037
Female Lunch  1.250  5.170     16.339        18.300
       Dinner 1.000  6.500     19.213        21.519

# definir filtros

lo hacemos creando una funcion

la funcion trasforma euros a dolares y de eso voy a escoger la columna total factura y le digo que calcule el promedio y que me diga todoas las factuas mayores a 20



In [ ]:
def f_filter(x):
  return mean_euro2usd(x['total_bill'].mean())>20

In [ ]:
df_filtered = df.groupby(['sex','time']).filter(f_filter)
df_filtered

total_bill   tip     sex smoker   day    time  size  prct_tip
0        16.990 1.010  Female     No   Sun  Dinner     2     0.059
1        10.340 1.660    Male     No   Sun  Dinner     3     0.161
2        21.010 3.500    Male     No   Sun  Dinner     3     0.167
3        23.680 3.310    Male     No   Sun  Dinner     2     0.140
4        24.590 3.610  Female     No   Sun  Dinner     4     0.147
..          ...   ...     ...    ...   ...     ...   ...       ...
239      29.030 5.920    Male     No   Sat  Dinner     3     0.204
240      27.180 2.000  Female    Yes   Sat  Dinner     2     0.074
241      22.670 2.000    Male    Yes   Sat  Dinner     2     0.088
242      17.820 1.750    Male     No   Sat  Dinner     2     0.098
243      18.780 3.000  Female     No  Thur  Dinner     2     0.160

[209 rows x 8 columns]

con el filtro anterior solo apareciron 209 elementos de los 243 , osea los que cumplen el filtro 

### aqui observamos que el filtro funcionao porque al pedir que cuente los datos  como los launch de mujeres estaban debajo de 20 entonces no pasaron al df_filtered

In [ ]:
df_filtered.groupby(['sex','time']).count()

total_bill     tip  smoker     day    size  prct_tip
sex    time                                                        
Male   Lunch       33.000  33.000  33.000  33.000  33.000    33.000
       Dinner     124.000 124.000 124.000 124.000 124.000   124.000
Female Lunch          nan     nan     nan     nan     nan       nan
       Dinner      52.000  52.000  52.000  52.000  52.000    52.000

# **22** Group By: extraer valor con variables categóricas


ahora vamos a extraer valor esta vez no usando las variables numericas si no tambien las mismas variables categoricas

In [14]:
df['ones']=1
df


total_bill   tip     sex smoker   day    time  size bin_total  ones
0        16.990 1.010  Female     No   Sun  Dinner     2   (3, 18]     1
1        10.340 1.660    Male     No   Sun  Dinner     3   (3, 18]     1
2        21.010 3.500    Male     No   Sun  Dinner     3  (18, 35]     1
3        23.680 3.310    Male     No   Sun  Dinner     2  (18, 35]     1
4        24.590 3.610  Female     No   Sun  Dinner     4  (18, 35]     1
..          ...   ...     ...    ...   ...     ...   ...       ...   ...
239      29.030 5.920    Male     No   Sat  Dinner     3  (18, 35]     1
240      27.180 2.000  Female    Yes   Sat  Dinner     2  (18, 35]     1
241      22.670 2.000    Male    Yes   Sat  Dinner     2  (18, 35]     1
242      17.820 1.750    Male     No   Sat  Dinner     2   (3, 18]     1
243      18.780 3.000  Female     No  Thur  Dinner     2  (18, 35]     1

[244 rows x 9 columns]

# y ahora voy a hacer grupos con estas categorias 

In [ ]:
df_g=df.groupby(['sex','smoker'])[['ones']].sum()
df_g

ones
sex    smoker      
Male   Yes       60
       No        97
Female Yes       33
       No        54

# vamos a agrupar por el nivel de los indices

que agrupe en el nivel cero es decir el genero 

In [ ]:
df_g.groupby(level=0).apply(
    
    lambda x :
    x/x.sum()*100
)

ones
sex    smoker       
Male   Yes    38.217
       No     61.783
Female Yes    37.931
       No     62.069

# otra forma de extraer valor de nuestros datos es transformar nuestras variables numericas en una variable categorica

lo vamos a hacer para convertirlo en una variable categorica para ciertos intervalos 

esto lo hacemos con cut

In [ ]:
pd.cut(df['total_bill'], bins =3)

# lo que sucede es que cada uno de los valores que estaba definido previamente
#va a quedar en una categoria 

0       (3.022, 18.983]
1       (3.022, 18.983]
2      (18.983, 34.897]
3      (18.983, 34.897]
4      (18.983, 34.897]
             ...       
239    (18.983, 34.897]
240    (18.983, 34.897]
241    (18.983, 34.897]
242     (3.022, 18.983]
243     (3.022, 18.983]
Name: total_bill, Length: 244, dtype: category
Categories (3, interval[float64]): [(3.022, 18.983] < (18.983, 34.897] < (34.897, 50.81]]

# Para ver las categorias que quedaron 

uso 

```
.value_counts()
```

In [ ]:
pd.cut(df['total_bill'], bins =3).value_counts()

(3.022, 18.983]     140
(18.983, 34.897]     88
(34.897, 50.81]      16
Name: total_bill, dtype: int64

# tambien podemos definir el ancho de nuestras categorias 


In [ ]:
pd.cut(df['total_bill'], bins =[3,18,35,60]).value_counts()

(3, 18]     127
(18, 35]    101
(35, 60]     16
Name: total_bill, dtype: int64

vamos a hacer un grupbay que tenga en cuenta el tiempo y la categoria que creamos bin_total

In [16]:
df['bin_total']=pd.cut(df['total_bill'], bins =[3,18,35,60])
df

total_bill   tip     sex smoker   day    time  size bin_total  ones
0        16.990 1.010  Female     No   Sun  Dinner     2   (3, 18]     1
1        10.340 1.660    Male     No   Sun  Dinner     3   (3, 18]     1
2        21.010 3.500    Male     No   Sun  Dinner     3  (18, 35]     1
3        23.680 3.310    Male     No   Sun  Dinner     2  (18, 35]     1
4        24.590 3.610  Female     No   Sun  Dinner     4  (18, 35]     1
..          ...   ...     ...    ...   ...     ...   ...       ...   ...
239      29.030 5.920    Male     No   Sat  Dinner     3  (18, 35]     1
240      27.180 2.000  Female    Yes   Sat  Dinner     2  (18, 35]     1
241      22.670 2.000    Male    Yes   Sat  Dinner     2  (18, 35]     1
242      17.820 1.750    Male     No   Sat  Dinner     2   (3, 18]     1
243      18.780 3.000  Female     No  Thur  Dinner     2  (18, 35]     1

[244 rows x 9 columns]

In [18]:
df.groupby(['time','bin_total'])[['ones']].count()

ones
time   bin_total      
Lunch  (3, 18]      45
       (18, 35]     21
       (35, 60]      2
Dinner (3, 18]      82
       (18, 35]     80
       (35, 60]     14

# ahora lo anterior en porcentajes

In [19]:
df.groupby(['time','bin_total'])[['ones']].count().groupby(level=0).apply(
    
    lambda x :
    x/x.sum()*100
)

ones
time   bin_total       
Lunch  (3, 18]   66.176
       (18, 35]  30.882
       (35, 60]   2.941
Dinner (3, 18]   46.591
       (18, 35]  45.455
       (35, 60]   7.955